### PR TITLE
WIP Design of new st2 ask feature

### PR DESCRIPTION
**Pull request by Mierdin**
- [ ] Provide a step-by-step (maybe diagram?) of functionality, end to end
- [x] Add example workflow with st2.ask in action

### CODE REVIEW

**Comment by m4dcoder** [comment]
What's the mechanism of this st2.ask action? How will it prompt user for inputs? How will it pause the workflow?

**Comment by enykeev** [comment]
> What's the mechanism of this st2.ask action?

st2.ask action is a dual-purpose one.

On the first run, it will check that the required data is missing in the execution context and return the instruction to its runner to transition status of this execution to awaits. At that point the execution will be considered complete and no other tasks from this workflow should be scheduled until it remains in awaits state. For mistral runner it should look mostly the same way as cancelled status.

User would see the execution in the `awaits` state (either by listening to stream or by periodically checking the list of executions) and would be able to manually get it back to `running` state by hitting new endpoint with the data requested by the st2.ask. This API call will trigger the rescheduling the st2.ask action.

On the second run, st2.ask will once again check the data in the context and could either return `succeeded` status resuming the workflow execution or `failed`/`awaits` status if it's still not satisfied with the data provided by the user.

> How will it prompt user for inputs?

The presence of `awaits` execution in the list by default implies the request for additional data. The UX have to be built on top of that. For st2web, you'll have a separate app for dealing with st2.ask requests. CLI would require you to once in a while check execution list. One can set up a rule to be notified about executions switching to awaits by email, slack or whatever. Since transition back from awaits requires just an API call, the `st2.answer` action could also be implemented to simplify things.

File: st2ask-planning.md [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+## Runners
+
+- [ ] Right now, the runner logic is pretty boolean true/false. Needs to be more dymanic. See [here](https://github.com/StackStorm/st2/blob/master/contrib/runners/python_runner/python_runner.py#L186) and [here](https://github.com/StackStorm/st2/blob/master/contrib/runners/python_runner/python_runner.py#L215) for some interesting bits of the code for the python runner.
+
+## API
+
+- [ ] Add an API endpoint like `PUT /execution/{id}` for user to request execution be put into a certain state. Support ability for client to pass arbitrary data, which will eventually get placed in the execution's context.
+- [ ] Implement the back-end API controller to pass this request to `st2common.services.action` and pass incoming data to execution's context
+- [ ] Add ability to request state change to `st2common.services.action`
+- [ ] Probably will need to add something to streaming API to notify clients of new `awaits` executions?
```
Comment (enykeev) [inline st2ask-planning.md]: st2stream already notifies listeners about any change in execution documents. Changing a status to awaits will automatically trigger `st2.execution__update` notification.

File: st2ask-planning.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+## Misc
+
+- [ ] Support new `awaits` status returned from Actions. I am assuming that this should also figure out the parent execution and also call out to Mistral (or ActionChain) to pause the workflow. TODO(mierdin): Look at what happens when an action returns "failed" since this probably does something similar.
+- [ ] Creation of `st2.ask` action that validates data and returns `awaits` if invalid, or success if valid.
+
+# Questions
+
+- The primary use case for this is in workflows. However, I'm not sure if actions within a workflow know if they were invoked by a workflow. Presumably, something would need to detect if a, for instance, `core.local` action returns `awaits`, and if so, send a message to Mistral to pause the workflow that action was a part of. Still looking, but don't know how/if this is currently possible.
+
+- Haven't really had a chance to think about what might consume the new `awaits` data from st2stream - something that might allow users to respond via email or slack to resume a workflow instead of using the CLI.
```
Comment (enykeev) [inline st2ask-planning.md]: We do already have action_notifier (or whatever it called, the trigger that happens on execution completion). No need to use st2stream for that unless you want notification to be handled outside of st2 (i.e. in sensu or whatever).

File: st2ask-planning.md [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+- Where is the best place to reach out to Mistral or ActionChain to pause the workflow? Originally I was thinking the Mistral or ActionChain runners but it wouldn't be an ActionChain or Mistral execution that would return an `awaits` status, it would be something like the Python runner. So not entirely sure where this happens just yet.
```
Comment (enykeev) [inline st2ask-planning.md]: Yeah, just go read the code on how the new execution get scheduled in response to the completion of previous one. I don't remember the details, but I'm reasonably certain you don't need to actively tell mistral or actionchain to pause, you just need to teach them not to schedule more actions when the previous one finished with awaits (see cancel, the same logic really).

**Comment by m4dcoder** [comment]
> st2.ask action is a dual-purpose one.
Which runner is the st2.ask action going to be running under?

**Comment by m4dcoder** [comment]
> One can set up a rule to be notified about executions switching to awaits by email, slack or whatever. 

I prefer that type of notification is defined at the time when the task is defined for st2.ask.  So instead of a separate rule, include who and how to notify (one or more) in the workflow as a parameter in st2.ask.

**Comment by m4dcoder** [comment]
> Since transition back from awaits requires just an API call, the st2.answer action could also be implemented to simplify things.

I see it as a matter of updating the st2.ask action execution entry over API and the "update" will trigger further action for this action execution, either to move resume the workflow and continue to wait.  What calls the API could be from the webUI or st2 CLI.  st2.answer may work but the action is in the user space and users can disable/modify the action.

**Comment by m4dcoder** [comment]
For the example(s), I like to see 
1. a decision made from st2.ask that either leads down path A or path B in the workflow
2. a multi-level dialog where the response of the st2.ask determines the next question to prompt the user

**Comment by m4dcoder** [comment]
What happens to the workflow execution if the user never responded to st2.ask? Another thing to consider. This is not critical in the next design iteration but needs to be addressed at some point.

**Comment by enykeev** [comment]
> Which runner is the st2.ask action going to be running under?

It might be any runner really, as long as we teach them to return a new status. Python one seems like an obvious choice for a start as it would give us enough flexibility to implement any scenario from the get go.

I'll leave the rest of the answers to Matt. :P

**Comment by m4dcoder** [comment]
> It might be any runner really, as long as we teach them to return a new status.

Something to consider... why not a new runner?  I don't like that the st2.action is in the user space and it'll be examining context and updating its own execution states. These should be done by the system.

**Comment by enykeev** [comment]
> why not a new runner?

Use cases where we would need slightly more complex logic (ex. 2fa) would require us to basically re-implement python runner inside this new runner. If we have a security concerns, we should probably add system level filtering and checks instead.

**Comment by m4dcoder** [comment]
> Use cases where we would need slightly more complex logic (ex. 2fa) would require us to basically re-implement python runner inside this new runner. If we have a security concerns, we should probably add system level filtering and checks instead.

I don't see a connection between the explanation here and a new runner for st2.ask.  Why do we need to re-implement python runner inside this new runner?  The way I see this is that the core logic of st2.ask will be implemented in this new runner.  The actions of this new runner type defines the parameters that get translated into the runner parameter (i.e. what to ask the user, what response is expected, how to notify the user, who to notify, etc.). Take a look at the expect runner as an example.

**Comment by m4dcoder** [comment]
And if we need to implement 2fa or other auth measures, more the reason this would be done in the system space instead of the user space.

**Comment by Mierdin** [comment]
Added example mistral workflow in https://github.com/Mierdin/st2-notes/pull/2/commits/931682bf3915bd6bbd2508b50b197cb556145b13

**Comment by Mierdin** [comment]
> What happens to the workflow execution if the user never responded to st2.ask? Another thing to consider. This is not critical in the next design iteration but needs to be addressed at some point.

I also am still considering this. Especially if the user provides their own JSON schema to `st2.ask`, I haven't thought of a way to figure out a "rejection" state, only a "not approved yet" state. So it would hang out in `awaits` indefinitely. It might be worth adding garbage collection logic somewhere. Maybe `st2garbagecollector`?? Totally don't know about that, haven't looked into it yet.

> why not a new runner?

Mirroring @enykeev comments here....the "approval" act may be simple binary yes/no but also may not be - and the latter can get pretty complicated. Having this in user space lets the user define their own validation in any language, instead of having them use our runner (which as @enykeev alludes, will probably mostly duplicate another already-existing runner). The potential complex nature of what "approval" could be, seems to indicate this needs to be done in a "serviceable" way. If `st2.ask` was any more complicated than "validate JSON and return status", I'd tend to agree with the runner approach so that logic is centralized. However, since it's really simple, I think it makes a bit more sense to leave this in user space and let it act as a reference implementation.

> (i.e. what to ask the user, what response is expected, how to notify the user, who to notify, etc.).

The current plan is to leave such things outside of `st2.ask`. The `st2.ask` action will only be responsible for blocking a workflow until the data stored in execution context matches the schema in use. These higher-level UX type of things can take place in a separate component that pushes data to execution API

> And if we need to implement 2fa or other auth measures, more the reason this would be done in the system space instead of the user space.

I can't think of a way to do that; the idea of doing this in user space is that the validation of this data is only possible by the user, since we can't/shouldn't anticipate every way this type of data should be represented. Let the user pass in their own JSON schema, or worst case, write their own version of `st2.ask` that has more detailed logic.

**Comment by Mierdin** [comment]
Btw I am still working through earlier comments, including adding a step-by-step, so hopefully things become clearer in the next few days. I published this last week very half-finished, in the interest of transparency

File: st2ask-planning.md (hunk 1 above)
Comment (Mierdin) [inline st2ask-planning.md]: Got it. So no anticipated changes here?

File: st2ask-planning.md (hunk 3 above)
Comment (Mierdin) [inline st2ask-planning.md]: Ack - thank you!

File: st2ask-planning.md (hunk 2 above)
Comment (Mierdin) [inline st2ask-planning.md]: Got it, that's helpful. I'll update the document with an FAQ section, and I'll address this there.

**Comment by m4dcoder** [comment]
> the "approval" act may be simple binary yes/no but also may not be - and the latter can get pretty complicated. Having this in user space lets the user define their own validation in any language, instead of having them use our runner

Please present a concrete use case for this complex approval with it's own validation in any language.

**Comment by m4dcoder** [comment]
Also, I have a strong opinion for a streamline and curated experience on how to define approval or other user interaction in the workflow. I agree that making different low level primitives and let users put them together is empowering. But from support experience, unless it's really clear. simple, and straightforward to put these together, it becomes a support burden.

**Comment by m4dcoder** [comment]
> (i.e. what to ask the user, what response is expected, how to notify the user, who to notify, etc.).

> The current plan is to leave such things outside of st2.ask

1. The name `st2.ask` implies that there is an intended audience for the questions being asked. To leave these outside of st2.ask doesn't make sense to me.
2. If we are handling these w/ rules, as workflow author, it's too much work. I don't want to switch back and forth between working on a workflow and jumping to work on rules and triggers. The UX is just going to be worst if someone uses the web UI for writing workflows and rules. Imagine jumping back and forth between st2flow and the rule designer.

**Comment by Mierdin** [comment]
So I agree with the UX-related drawbacks of the approach - but to me it seems like it's a tradeoff between that and creating a lot of duplicate functionality. This way we're just using rules in the way they should be used, to respond to an event (change in execution status) and pass a message to a user with an action (slack, etc), instead of building in any form of communication (email, slack, etc) to a single action *as well as* the logic for listening for responses. Seems like that's a wasted effort.
So if this is too complicated, I'm all for going a different route but I'm having a hard time coming up with alternatives that don't reinvent the wheel; maybe I'm thinking about it wrong. If there's a simpler way that I'm missing, please point me in that direction

On a related but separate note, if renaming `st2.ask` will make it more obvious what it's doing, I'm all for that.

**Comment by m4dcoder** [comment]
> I'm curious on your thoughts re: requesting user input in an `st2.ask` action. The model I'm currently working off of allows the action to return more or less immediately, so nothing's long-running, just will resume after API call. Do you envision `st2.ask` sending slack or email message, and listen for response somehow?

So, this is from my line of thoughts on implementing it as runner...  the st2.ask runner will put the action in a `pending` state, then depending on the types of notification, schedules appropriate actions to do the notification (i.e. email, slack, etc.), set the parent workflow execution to pause, and then the runner exit. the user responds over webUI, CLI, etc..  the execution result gets updated w/ the response over API and the API controller moves the action back to a scheduled state. This will route the execution back to the runner. The runner inspects the result and if the response is satisfactory, then mark the execution as complete, resumes the parent workflow, then exit.

Side notes:
- If we really want to use rules, then in the runner, send the webhooks to trigger the notifications instead of calling the notification actions directly.  They payload of the webhook should contain the list of users to notify. This ensures that the list of who to notify is defined as part of the task in the workflow. The user can enable/disable the rules to control which notification type to use. 
- If we are implementing this as a runner, the entry point of the action will point to the specs for the questions being asked.

**Comment by m4dcoder** [comment]
@Mierdin, If you are not convinced, for comparison, you can try to come up with similar solution as above using st2.ask as an action and the rules that you are thinking about.

**Comment by Mierdin** [comment]
I wasn't aware that runners/actions could schedule other actions. Is there an example I can use for reference? Also, if the `st2.ask` action calls out to schedule a notification action, I'm assuming we would mandate that the user already has those packs configured? That way we can just refer to the communication method in the `st2.ask` params.

And thanks for generally adding more re: your line of thinking, I was right, I overthought it - we're more on the same page than I thought.

**Comment by m4dcoder** [comment]
Runner is a system process and has access to common services, utils, and DB. And so if we write a new runner, we can have it schedule new actions or push webhooks. There's no example runner. 

Action on the other hand is in the user space and I don't recommend it to execute other actions. This sets a bad precedent.

**Comment by Mierdin** [comment]
This changes things for me. Thanks @m4dcoder will ponder in the car and modify documents accordingly.

File: st2ask-planning.md (hunk 3 above)
Comment (Mierdin) [inline st2ask-planning.md]: So I did some code walking tonight and basically what happens is that the liveaction_db and execution objects have their statuses updated and written to the DB, then the liveaction_ob instance  is sent out on the message queue. I'm still looking, but I'm assuming at this point I'll need to look for somewhere else that's watching for either of these things and instructing Mistral not to schedule more events....perhaps it's mistral itself that's doing this. Still looking

Comment (Mierdin) [inline st2ask-planning.md]: Oh, duh - this is what st2resultstracker is for

**Pull request closed without merging**